File: apps/ios/BitOS/Platform/Author/ProfileLookupStore.swift
import BusinessCore
import Foundation
import Observation

/**
 * One-shot multi-pubkey kind-0 lookup (legacy `SupportController` /
 * `ContributorsWidget` fetch parity): subscribes `kinds:[0] authors:[…]`
 * through the shared codec, keeps the NEWEST verified metadata per pubkey,
 * and settles after 8 s. Pure display projection; no cache, no persistence.
 */
@MainActor
@Observable
final class ProfileLookupStore {
    private(set) var profiles: [String: ProfileMetadata] = [:]
    private(set) var loading = false

    private let pool: RelayPool
    private let client: any BusinessCoreClient
    private var seenAt: [String: Int64] = [:]
    private var wanted: Set<String> = []
    private var settleTask: Task<Void, Never>?
    private var framesTask: Task<Void, Never>?

    init(pool: RelayPool, client: any BusinessCoreClient) {
        self.pool = pool
        self.client = client
    }

    func start() {
        guard framesTask == nil else { return }
        Task { [weak self, pool, client] in
            let stream = await pool.verifiedFrames(client: client)
            guard let self, !Task.isCancelled else { return }
            self.framesTask = FrameIngest.pump(
                gated: stream,
                isAlive: { [weak self] in self != nil }
            ) { [weak self] gated in
                await self?.absorb(gated)
            }
        }
    }

    private func absorb(_ gated: GatedFrame) {
        guard case .event(let gatedEvent) = gated,
              gatedEvent.event.kind == 0,
              wanted.contains(gatedEvent.event.pubkey) else { return }
        let decoded = gatedEvent.event
        let known = seenAt[decoded.pubkey] ?? Int64.min
        if decoded.createdAt < known { return }
        guard let metadata = client.profile(from: decoded) else { return }
        seenAt[decoded.pubkey] = decoded.createdAt
        profiles[decoded.pubkey] = metadata
    }

    /// Looks up one or more pubkeys; settles (loading=false) after 8 s.
    func lookup(pubkeys: [String]) {
        wanted = Set(pubkeys)
        seenAt = [:]
        profiles = [:]
        loading = true
        settleTask?.cancel()
        settleTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(8))
            self?.loading = false
        }
        let request = client.profileRequest(subscriptionId: "bitos-lookup", authors: pubkeys)
        guard !request.isEmpty else { return }
        Task { await pool.broadcast(request) }
    }

    func stop() {
        wanted = []
        loading = false
        settleTask?.cancel()
    }
}

File: apps/android/src/main/kotlin/space/bitos/app/data/feed/ProfileLookupStore.kt
package space.bitos.app.data.feed

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import space.bitos.app.data.relay.RelayPool
import space.bitos.app.data.relay.VerifiedPoolFrame
import space.bitos.core.model.ProfileMetadata
import space.bitos.core.nostr.EventHasher
import space.bitos.core.nostr.NostrEventCodec
import space.bitos.core.nostr.Sha256EventHasher

/**
 * One-shot multi-pubkey kind-0 lookup (legacy `SupportController` /
 * `ContributorsWidget` fetch parity): subscribes `kinds:[0] authors:[…]`,
 * keeps the NEWEST verified metadata per pubkey, and settles after a
 * timeout. Pure display projection; no cache, no persistence.
 */
class ProfileLookupStore(
    private val scope: CoroutineScope,
    private val pool: RelayPool,
    private val hasher: EventHasher = Sha256EventHasher,
) {
    private var collectJob: Job? = null
    private var settleJob: Job? = null

    private val _profiles = MutableStateFlow<Map<String, ProfileMetadata>>(emptyMap())
    val profiles: StateFlow<Map<String, ProfileMetadata>> = _profiles.asStateFlow()

    private val _loading = MutableStateFlow(false)
    val loading: StateFlow<Boolean> = _loading.asStateFlow()

    private var seenAt = mutableMapOf<String, Long>()
    private var wanted: Set<String> = emptySet()

    init {
        collectJob = scope.launch {
            pool.verifiedFrames.collect { gated ->
                val event = (gated as? VerifiedPoolFrame.Verified)?.event ?: return@collect
                if (event.kind != space.bitos.core.model.NostrKinds.PROFILE_METADATA) return@collect
                val key = event.pubkey.value
                if (key !in wanted) return@collect
                val known = seenAt[key] ?: Long.MIN_VALUE
                if (event.createdAt < known) return@collect
                val metadata = ProfileMetadata.parse(event) ?: return@collect
                seenAt[key] = event.createdAt
                _profiles.value = _profiles.value + (key to metadata)
            }
        }
    }

    /** Looks up one or more pubkeys; settles (loading=false) after 8 s. */
    fun lookup(pubkeys: List<String>) {
        wanted = pubkeys.toSet()
        seenAt = mutableMapOf()
        _profiles.value = emptyMap()
        _loading.value = true
        settleJob?.cancel()
        settleJob = scope.launch {
            delay(8_000)
            _loading.value = false
        }
        val authors = pubkeys.joinToString(prefix = "[\"", separator = "\",\"", postfix = "\"]")
        pool.broadcast(
            NostrEventCodec.encodeRequest(
                "bitos-lookup",
                "{\"kinds\":[0],\"authors\":$authors,\"limit\":${(pubkeys.size * 2).coerceAtMost(20)}}",
            ),
        )
    }

    fun stop() {
        wanted = emptySet()
        _loading.value = false
        settleJob?.cancel()
    }
}
